new-cep.md

```diff
@@ -0,0 +1,218 @@
+behavior, we propose a set of generic plugin hooks in this CEP. Included will 
+be `pre_command` and `post_command` hooks that will allow
+plugin authors to execute their plugin code before or after conda commands
+run, respectively. Additionally, we introduce the `exception_handler` plugin 
+hook that will allow plugin authors to override and extend behavior of the
+existing [`conda.exceptions.ExceptionHandler`][conda-exception-handler] class. 
+For each hook, we outline example use cases and
+show exactly how plugin authors will define these new hooks.
+
+## Specification
+
+This specification calls for the creation of the following three new plugin hooks:
+
+- `pre_command`: runs before the invoked conda command is run
+- `post_command`: runs after the invoked conda command is run
+- `exception_handler`: allows overriding or extending the existing
+  [`conda.exceptions.ExceptionHandler`][conda-exception-handler]
+
+Below, we discuss an example showing the use of the `pre_command` and `post_command` 
+hooks together and an example of the `exception_handler` hook.
+
+### `pre_command` and `post_command`
+
+Plugin authors create the `pre_command` and `post_command` hooks by first defining a `conda.plugins.hookimpl`
+decorated function called either `conda_pre_command` or `conda_post_command` which return either a `CondaPreCommand` 
+or `CondaPostCommand` class, respectively. These classes will both receive the following three properties:
+
+- `name`: unique name which identifies this plugin hook
+- `action`: a callable which contains the code to be run
+- `run_for`: a Python `set` of strings representing the commands this will be run on (e.g. `install` and `create`)
+
+#### Example
+
+```python
+from conda.plugins import hookimpl, CondaPreCommand, CondaPostCommand
+
+
+PLUGIN_NAME = "custom_plugin"
+
+
+def custom_plugin_pre_command_action(command):
+    """
+    Defines our custom pre-run action which simply prints a message.
+    """
+    print(f"pre-run action before {command}")
+
+
+@hookimpl
+def conda_pre_commands():
+    """
+    Yields our CondaPreCommand instance which attaches our ``custom_plugin_pre_command_action``
+    to the "install" and "create" command.
+    """
+    yield CondaPreCommand(
+        name=f"{PLUGIN_NAME}_pre_command",
+        action=custom_plugin_pre_command_action,
+        run_for={"install", "create"}
+    )
+
+
+def custom_plugin_post_command_action(command):
+    """
+    Defines our custom post-run action which simply prints a message.
+    """
+    print(f"post-run action after {command}")
+
+
+@hookimpl
+def conda_post_commands():
+    """
+    Yields our CondaPreCommand instance which attaches our ``custom_plugin_post_command_action`` to
+    the "install" and "create" command.
+    """
+    yield CondaPostCommand(
+        name=f"{PLUGIN_NAME}_post_command",
+        action=custom_plugin_post_command_action,
+        run_for={"install", "create"}
+    )
+```
+
+### `exception_handler`
```

Comment:
can we review, if current implementation of exception classes should be extended to support custom exception handlers? For example - let's say in the code we have (just found the first one)
```
    if args.all and prefix == context.default_prefix:
        msg = "cannot remove current environment. deactivate and run conda remove again"
        raise CondaEnvironmentError(msg)
```

If I want to write a custom exception handler for his exception, I'm missing context - which environment and reason, why it couldn't be removed (or is conda Context the right thing to use?)

If quick question - yes, it worth to review and extend exception handling mechanism - I believe it's another CEP?

Reply:
@aovasylenko, please see @jaimergp's comment below. I think he did a reasonable job of summarizing what you were talking about.

The scenario you are talking about would require a very extensive refactor to existing code, and I do not think it is realistic anytime soon. In the meantime, it's way better to just rely on the context object like you said.